crates/services/p2p/src/service.rs

```diff
@@ -858,14 +1000,85 @@ impl SharedState {
         let (response_from_peer, response) =
             receiver.await.map_err(|e| anyhow!("{e}"))?;
         assert_eq!(
-            peer_id,
+            peer_id.as_ref(),
             response_from_peer.to_bytes(),
             "Bug: response from non-requested peer"
         );
 
         response.map_err(|e| anyhow!("Invalid response from peer {e:?}"))
     }
 
+    pub async fn get_all_transactions_ids_from_peer(
+        &self,
+        peer_id: FuelPeerId,
+    ) -> anyhow::Result<Vec<TxId>> {
+        let (sender, receiver) = oneshot::channel();
+        let from_peer = PeerId::from_bytes(peer_id.as_ref()).expect("Valid PeerId");
+        let request = TaskRequest::TxPoolGetAllTxIds {
+            from_peer,
+            channel: sender,
+        };
+        self.request_sender.try_send(request)?;
+
+        let (response_from_peer, response) =
+            receiver.await.map_err(|e| anyhow!("{e}"))?;
+
+        debug_assert_eq!(
+            peer_id.as_ref(),
+            response_from_peer.to_bytes(),
+            "Bug: response from non-requested peer"
+        );
+
+        let Some(txs) =
+            response.map_err(|e| anyhow!("Invalid response from peer {e:?}"))?
+        else {
+            return Ok(vec![]);
+        };
+        if txs.len() > self.max_txs_per_request {
+            return Err(anyhow!("Too many transactions requested: {}", txs.len()));
+        }
+        Ok(txs)
+    }
+
+    pub async fn get_full_transactions_from_peer(
+        &self,
+        peer_id: FuelPeerId,
+        tx_ids: Vec<TxId>,
+    ) -> anyhow::Result<Vec<Option<Transaction>>> {
+        let (sender, receiver) = oneshot::channel();
+        let from_peer = PeerId::from_bytes(peer_id.as_ref()).expect("Valid PeerId");
+        let request = TaskRequest::TxPoolGetFullTransactions {
+            tx_ids,
+            from_peer,
+            channel: sender,
+        };
+        self.request_sender.try_send(request)?;
+
+        let (response_from_peer, response) =
+            receiver.await.map_err(|e| anyhow!("{e}"))?;
+        debug_assert_eq!(
+            peer_id.as_ref(),
+            response_from_peer.to_bytes(),
+            "Bug: response from non-requested peer"
+        );
+
+        let Some(txs) =
+            response.map_err(|e| anyhow!("Invalid response from peer {e:?}"))?
+        else {
+            return Ok(vec![]);
+        };
+        if txs.len() > self.max_txs_per_request {
```

Comment:
Maybe we need to not send more than `self.max_txs_per_request` as well=)

Reply:
Done

Reply:
I meant it would be nice from TxPool not to request more than our own config allows=)

Reply:
Oopsie, Changed.